spikeextractors/extractors/nwbextractors/nwbextractors.py

```diff
@@ -501,81 +497,111 @@ def add_epochs(recording, nwbfile):
         '''
         assert HAVE_NWB, NwbRecordingExtractor.installation_mesg
 
         assert distutils.version.LooseVersion(pynwb.__version__) >= '1.3.3', \
             "'write_recording' not supported for version < 1.3.3. Run pip install --upgrade pynwb"
-
-        if os.path.exists(save_path):
-            read_mode = 'r+'
-        else:
-            read_mode = 'w'
-
+            
+        assert save_path is None or nwbfile is None, \
+            'Either pass a save_path location, or nwbfile object, but not both!'
+            
         # Update any previous metadata with user passed dictionary
         if metadata is None:
             metadata = dict()
         if hasattr(recording, 'nwb_metadata'):
             metadata = update_dict(recording.nwb_metadata, metadata)
 
-        with NWBHDF5IO(save_path, mode=read_mode) as io:
-            if read_mode == 'r+':
-                nwbfile = io.read()
+        if nwbfile is None:
+            if os.path.exists(save_path):
+                read_mode = 'r+'
             else:
-                if 'NWBFile' not in metadata:
-                    metadata['NWBFile'] = {'session_description': 'no description',
-                                           'identifier': str(uuid.uuid4()),
-                                           'session_start_time': datetime.now()}
-                nwbfile = NWBFile(**metadata['NWBFile'])
-
-            # Add devices
-            nwbfile = se.NwbRecordingExtractor.add_devices(
-                recording=recording,
-                nwbfile=nwbfile,
-                metadata=metadata
-            )
-
-            # Add electrode groups
-            nwbfile = se.NwbRecordingExtractor.add_electrode_groups(
-                recording=recording,
-                nwbfile=nwbfile,
-                metadata=metadata
-            )
-
-            # Add electrodes
-            nwbfile = se.NwbRecordingExtractor.add_electrodes(
-                recording=recording,
-                nwbfile=nwbfile,
-                metadata=metadata
-            )
-
-            # Add electrical series
-            nwbfile = se.NwbRecordingExtractor.add_electrical_series(
-                recording=recording,
-                nwbfile=nwbfile,
-                metadata=metadata
-            )
-
-            # Add epochs
-            nwbfile = se.NwbRecordingExtractor.add_epochs(
-                recording=recording,
-                nwbfile=nwbfile
+                read_mode = 'w'
+    
+            with NWBHDF5IO(save_path, mode=read_mode) as io:
+                if read_mode == 'r+':
+                    nwbfile = io.read()
+                else:
+                    if 'NWBFile' not in metadata:
+                        metadata['NWBFile'] = {'session_description': 'no description',
+                                               'identifier': str(uuid.uuid4()),
+                                               'session_start_time': datetime.now()}
```

Comment:
this should use `update_dict` to handle if metadata['NWBFile'] is partially defined

Reply:
Good call!